File: backend/src/main/java/ai/alvoff/research/ResearchClient.java
package ai.alvoff.research;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;

@Component
@RequiredArgsConstructor
@Slf4j
public class ResearchClient {

    private final ObjectMapper mapper;

    @Value("${app.research-agent.url}")
    private String agentUrl;

    private final HttpClient client = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(15))
        .build();

    /**
     * Stream SSE events from the Python agent. The consumer is invoked once per
     * complete event with (eventName, jsonDataString). Returns when the stream
     * closes (after `done` or `error`, or on connection close).
     */
    public void streamResearch(
        String query,
        List<Map<String, String>> history,
        BiConsumer<String, String> onEvent
    ) throws Exception {
        Map<String, Object> body = Map.of(
            "query", query,
            "conversation_history", history
        );
        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create(agentUrl + "/research"))
            .timeout(Duration.ofMinutes(10))
            .header("Content-Type", "application/json")
            .header("Accept", "text/event-stream")
            .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body)))
            .build();

        HttpResponse<InputStream> resp = client.send(req, HttpResponse.BodyHandlers.ofInputStream());
        if (resp.statusCode() / 100 != 2) {
            String err = new String(resp.body().readAllBytes(), StandardCharsets.UTF_8);
            throw new RuntimeException("Research agent returned " + resp.statusCode() + ": " + err);
        }

        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(resp.body(), StandardCharsets.UTF_8))) {

            String currentEvent = null;
            StringBuilder dataBuf = new StringBuilder();
            String line;

            while ((line = reader.readLine()) != null) {
                if (line.isEmpty()) {
                    if (currentEvent != null && dataBuf.length() > 0) {
                        try {
                            onEvent.accept(currentEvent, dataBuf.toString());
                        } catch (Exception consumerErr) {
                            log.warn("SSE consumer threw: {}", consumerErr.getMessage());
                        }
                    }
                    currentEvent = null;
                    dataBuf.setLength(0);
                } else if (line.startsWith("event: ")) {
                    currentEvent = line.substring(7).trim();
                } else if (line.startsWith("data: ")) {
                    if (dataBuf.length() > 0) dataBuf.append("\n");
                    dataBuf.append(line.substring(6));
                }
                // ignore comments (": ...") and other lines
            }
        }
    }
}

File: backend/app/research_client.py
"""Streams SSE events from the research-agent service."""
from collections.abc import AsyncIterator

import httpx

from .config import settings


class AgentEvent:
    __slots__ = ("name", "data")

    def __init__(self, name: str, data: str) -> None:
        self.name = name
        self.data = data


async def stream_research(
    query: str, history: list[dict[str, str]]
) -> AsyncIterator[AgentEvent]:
    """Yield each complete SSE event from the agent (event:/data: pair)."""
    body = {"query": query, "conversation_history": history}
    timeout = httpx.Timeout(connect=15.0, read=600.0, write=30.0, pool=15.0)

    async with httpx.AsyncClient(timeout=timeout) as client:
        async with client.stream(
            "POST",
            f"{settings.research_agent_url}/research",
            json=body,
            headers={"Accept": "text/event-stream"},
        ) as resp:
            if resp.status_code // 100 != 2:
                body_text = await resp.aread()
                raise RuntimeError(
                    f"Research agent returned {resp.status_code}: {body_text.decode(errors='replace')}"
                )

            current_event: str | None = None
            data_lines: list[str] = []

            async for raw_line in resp.aiter_lines():
                if raw_line == "":
                    if current_event and data_lines:
                        yield AgentEvent(current_event, "\n".join(data_lines))
                    current_event = None
                    data_lines = []
                elif raw_line.startswith("event: "):
                    current_event = raw_line[7:].strip()
                elif raw_line.startswith("data: "):
                    data_lines.append(raw_line[6:])
                # ignore comments and other lines

            if current_event and data_lines:
                yield AgentEvent(current_event, "\n".join(data_lines))
